[MoM] Wary Sleeper perk

Kray4000 opened the pull request:
#### Summary

Mods "[MoM] Add possible Wary Sleeper perk"

#### Purpose of change

Thought of a cool Clairsentience Perk.

#### Describe the solution

Should allow character with psinonic clairsentience powers to wake up when - variants:

- works like bio_alarm

or/and

- wake up character when near by NPCs who are currently hostile and creatures with aggression 10 or greater **but** then this perk must have possibility of turning off like active mutations, allows player to sleep on own risk 

Still need to:

- [x] Add `"enchantments"` line with `bio_alarm` to Wary Sleeper perk in perks.json

and 

- [x] Test this perk

#### Describe alternatives you've considered

Ask someone to add this perk

#### Testing

After complete previous tasks

#### Additional context

Will be appreciative for any help

Kray4000 committed:
[MoM] Wary Sleeper perk *initial commit

Add workpieces to next commit updates

Kray4000 commented:
> Something like this
Is that hard coded flag `MOTION_ALARM` for EOC?

GuardianDll commented:
Nope, it adds MOTION_ALARM enchantment, then you just add this enchantment to bionic when it's active (value 5) and use it for your trait

GuardianDll commented:
no, why, just make the trait activateable

Kray4000 commented:
All activateable perks have EOCs at least at Bombastic Perks files. Like you activate perk ok - run this EOC, deactivated? - This

GuardianDll commented:
trait, perk and mutations are all the same thing under the hood, `type: mutation`
check how ANTENNAE mutation works, it is activateable, but it doesn't fire eoc, it just signalizes the game that enchantment of `"condition": "ACTIVE"` can start work when it is active

Kray4000 committed:
.json, and .md .cpp .h update

*.perks.json update
*perkmenu.json update
*remove wary_sleeper.json
*add 991 line to Magic.md
*change in do_turn_cpp
*add 172 line in magic_enchantment.cpp
*add 148 line in magic_enchantment.h

Kray4000 commented:
Ok, I pushed changes (not tested yet), have spell_checker dictionary alert(don't know should I ignore it or add Wary Sleeper to list). How to test changes - run spell_check.py or spell_check_pr.py?

Kray4000 committed:
Validation fix

Kray4000 committed:
Astyle fix

Copy-Paste method

GuardianDll commented:
just add `premotion` to dictionary, as the bot suggests

Kray4000 committed:
miswrite fix (and astyle fix v.2)

Kray4000 committed:
Astyle fix 3

GuardianDll commented:
don't bother astyling, astyler bot gonna fix it when you undraft the PR

Kray4000 committed:
Add premotion to dictionary

GuardianDll commented:
but yeah you also have two json syntax errors

Kray4000 commented:
I forgot } and ] isn't it? in perks.json?

Kray4000 committed:
deleting comma

GuardianDll commented:
```Error parsing data/mods/MindOverMatter/mod_interactions/bombastic_perks/perks.json: Expecting ',' delimiter: line 80 column 3 (char 3548)```

Kray4000 committed:
adding comma

Kray4000 committed:
add #unclude "magic enchantment.h" to do_turn.cpp

Trying to fix IWYU error

Kray4000 committed:
#include 

Don't know why it UN if I was just copy paste that line from prev

Kray4000 commented:
And my job is done here(joke). 
`Error with General build matrix:  Error: src/do_turn.cpp:42:10: fatal error: 'magic enchantment.h' file not found` - ok added but then need to write magic_enchantment.h file, and I don't know how it works so can't do this.

GuardianDll commented:
well it would work if you included `magic_enchantment.h`, but you tried to include nonexistent `magic enchantment.h` instead

Kray4000 committed:
add ”_”

GuardianDll committed:
Update src/do_turn.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Kray4000 commented:
> You missed to add enchantment to bio_alarm
As mode override or just add enchantment to DDA item?

Kray4000 committed:
Add enchantment

added enchantment MOTION_ALARM to bio_alarm

Kray4000 committed:
Merge branch 'MoM-Wary-Sleeper' of https://github.com/Kray4000/Cataclysm-DDA into MoM-Wary-Sleeper

Kray4000 committed:
Revert "Merge branch 'MoM-Wary-Sleeper' of https://github.com/Kray4000/Cataclysm-DDA into MoM-Wary-Sleeper"

This reverts commit bff6b10f41d69f2034f7842bf900f9d948023e6c, reversing
changes made to 4fb0995bcacf97e7915c81a3febcae25ac2ac18a.

Kray4000 committed:
Update src/do_turn.cpp

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Kray4000 committed:
Update perkmenu.json "-,"

Kray4000 committed:
Update do_turn.cpp 

#include "magic_enchantment.h"

Kray4000 committed:
Update perks.json

Kray4000 committed:
Update data/mods/MindOverMatter/mod_interactions/bombastic_perks/perks.json

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

Standing-Storm commented on data/mods/MindOverMatter/mod_interactions/bombastic_perks/perks.json:
```diff
@@ -67,6 +67,17 @@
+    "points": 0,
+    "purifiable": false,
+    "valid": false,
+    "description": "Your premotion abilities has developed to such a level that you sense movement around you even when you sleep.",
```
```suggestion
    "description": "Your Premonition ability has developed to such a level that you can sense movement around you even when you sleep.",
```

Standing-Storm commented on data/mods/MindOverMatter/mod_interactions/bombastic_perks/perks.json:
```diff
@@ -67,6 +67,17 @@
+    "purifiable": false,
+    "valid": false,
+    "description": "Your premotion abilities has developed to such a level that you sense movement around you even when you sleep.",
+    "active": true,
```
Is there a reason for it to be active (and have the active condition) instead of the condition being `{ "u_has_effect": "sleep" }` (which should make it automatically work when you fall asleep and turn off when you wake up, so you don't have to worry about activating it?

Kray4000 replied:
I'm not sure how it work, but this perk should have switch which you can turn off, and sleep on own risk, (it work like bio_alarm, that psionic copy-form) but if you don't agree with this, you can change it

GuardianDll replied:
It is active in case you indeed want to sleep in a hostile environment, it is 10 tiles range around character ignoring walls after all

Standing-Storm replied:
If the alarm automatically wakes you up, then yeah, it should definitely have a toggle for that reason

Kray4000 commented:
Ok, seems all checks passed, how to test changes on local game, with json it's easy, I'll just throw new files in my game folder what with .src? I tried to use spellchecker.py but seems it do nothing.

GuardianDll commented:
You should complie the game, by installing compiler

GuardianDll committed:
Update data/mods/MindOverMatter/mod_interactions/bombastic_perks/perks.json

Co-authored-by: Standing-Storm <120433252+Standing-Storm@users.noreply.github.com>

Kray4000 commented:
Ok, I tried to figure out how compilation works, read those instructions, download MSYS and then visual studio dev container. But then I realised that I can't understand how all that things work, it is too complicated for me, so I can't test this PR. Should I close it then?

GuardianDll commented:
I'll compile and test it in a few days myself

GuardianDll commented:
![image](https://github.com/user-attachments/assets/93d03d19-7d10-4e7f-be90-cb63080774a9)
seems to work as expected
